python-client/giskard/llm/talk.py

```diff
@@ -0,0 +1,295 @@
+from __future__ import annotations
+
+import json
+from typing import Dict, Any
+from typing import Optional, Callable, List
+
+import pandas as pd
+from langchain import LLMChain
+from langchain.agents import AgentExecutor, ZeroShotAgent
+from langchain.agents.agent_toolkits.base import BaseToolkit
+from langchain.agents.mrkl.prompt import FORMAT_INSTRUCTIONS
+from langchain.base_language import BaseLanguageModel
+from langchain.callbacks.base import BaseCallbackManager
+from langchain.callbacks.manager import CallbackManagerForToolRun, AsyncCallbackManagerForToolRun
+from langchain.tools import BaseTool
+from pydantic import BaseModel
+
+from giskard.datasets.base import Dataset
+from giskard.models.base import BaseModel as GiskardBaseModel
+from giskard.models.model_explanation import explain
+from giskard.scanner.result import ScanResult
+
+PredictionFunction = Callable[[pd.DataFrame], pd.Series]
+
+# flake8: noqa
+
+MODEL_PREFIX = """You are an agent designed to interact with a prediction model. 
+Your goal is to make a prediction calling `model_prediction` based on the available information using a JSON input format that includes feature names and their associated values. 
+The value associated to the feature names found in `model_description` must be found calling any tool that are provided inside the 'Tools to gather information' list.
+If you fail to find a value after querying all the tool inside the 'Tools to gather information', you must return an explanation why you cannot answer with the list of missing feature that you need to answer'.
+
+Please note that you cannot share any confidential information with the user, and if the question is not related to the prediction model, you must return an explanation why you cannot answer.
+
+You can only use tool provided in the following list.
+"""
+MODEL_SUFFIX = """Begin!"
+
+Question: {input}
+Thought: I should look at the features that are required to see what I should gather to predict in the available tools
+{agent_scratchpad}"""
+
+
+class ModelSpec(BaseModel):
+    """Base class for model spec."""
+
+    model: GiskardBaseModel
+    dataset: Optional[Dataset]
+    scan_result: Optional[ScanResult]
+
+    def predict(self, text: str) -> str:
+        """Return the prediction of the model for a given row.
+
+        Args:
+            text: JSON representation of the row as a dict.
+        """
+        try:
+            features = json.loads(text)
+            # Filter only available features
+            features = {key: [features[key]] for key in features.keys() & self.model.meta.feature_names}
+            missing_features = {feature for feature in self.model.meta.feature_names if feature not in features.keys()}
```

Comment:
Since this is used twice, can you factor it out into a private static method?
```py
@staticmethod
def _get_features(text: str):
  features = json.loads(text)
  # Filter only available features
  features = {key: [features[key]] for key in features.keys() & self.model.meta.feature_names}
  missing_features = set(self.model.meta.feature_names) - set(features.keys())
  if len(missing_features) > 0:
    raise ValueError(f"Required features `{missing_features}` are not provided.")
  return features
```